### PR TITLE
Country test fixture

### PR DESCRIPTION
**Pull request by blackbaud-svossen**
This PR will help addresses an issue (53) on the phone field component.

- Added a country field test fixture
- Added SkySelectTestingModule and SkyCountryFieldTestingModule to provide hooks into consumer's test modules, allowing us to do things like turn off animations as needed.

### CODE REVIEW

**Comment by blackbaud-ado** [comment]
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-country-test-fixture](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-country-test-fixture)

**Comment by blackbaud-ado** [comment]
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-country-test-fixture](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-country-test-fixture)

File: src/app/public/testing/country-field/country-field-fixture.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+   */
+  public get countryFlagIsVisible(): boolean {
+    const flag = this.getCountryFlag();
+    return flag !== null;
+  }
+
+  /**
+   * A flag indicating whether or not the input has been disabled.
+   */
+  public get isDisabled(): boolean {
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/testing/country-field/country-field-fixture.ts:64]: Could we rename this `disabled` to match the input property's name?

File: src/app/public/testing/country-field/country-field-fixture.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        .intlTelInputGlobals
+        .getCountryData()
+      )
+    );
+  }
+
+  /**
+   * The value of the input field's autocomplete attribute.
+   */
+  public get autocomplete(): string {
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/testing/country-field/country-field-fixture.ts:46]: Could we rename this `autocompleteAttribute` to match the component's property name? We have an autocomplete component and it comes with some predefined meanings of what "autocomplete" means, so we explicitly name this `autocompleteAttribute` to convey it controls the actual HTML attribute and not a SKYUX component.

File: src/app/public/testing/country-field/country-field-fixture.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  /**
+   * The value of the input field.
+   */
+  public get searchText(): string {
+    const inputEl = this.getInput();
+    return inputEl.nativeElement.value;
+  }
+
+  /**
+   * The value of the input field.
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/testing/country-field/country-field-fixture.ts:78]: This comment and the prior one are the same - could you expand to differentiate the two?

File: src/app/public/testing/country-field/country-field-fixture.ts [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    return this.fixture.whenStable();
+  }
+
+  //#region helpers
+
+  private getCountryFlag(): DebugElement {
+    return this.debugEl.query(By.css('.sky-country-field-flag'));
+  }
+
+  private getInput(): DebugElement {
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/testing/country-field/country-field-fixture.ts:119]: It looks like everywhere we're using this method, we're always converting it into the `nativeElement`. Should we scrap this method and just use `getInputElement()` below?

File: src/app/public/testing/country-field/country-field-fixture.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+  private enterSearch(newValue: string, fixture: ComponentFixture<any>): void {
+    const inputElement = this.getInputElement();
+    inputElement.value = newValue;
+
+    SkyAppTestUtility.fireDomEvent(inputElement, 'keyup');
+    fixture.detectChanges();
+    tick();
+    fixture.detectChanges();
+    tick();
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/testing/country-field/country-field-fixture.ts:145]: Are these `tick()`s doing anything? I thought they only are used when inside Angular's `fakeAsync` zone. We shouldn't need them here. If so, we may need to figure out a better way to do this.

File: src/app/public/testing/country-field/country-field-fixture.ts [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+   */
+  public get selectedCountryData(): SkyCountryFieldCountry {
+    return this.countries.find((x: any) => x.name === this.searchText);
+  }
+
+  /**
+   * Enters the search text into the input field displaying search results, but making no selection.
+   * @param searchText The name of the country to select.
+   */
+  public async search(searchText: string): Promise<NodeListOf<HTMLElement>> {
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/testing/country-field/country-field-fixture.ts:88]: Is this function actually asynchronous? If not, it doesn't need to return a promise.

File: src/app/public/testing/country-field/country-field-fixture.spec.ts [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+import {
+  SkyCountryFieldTestingModule
+} from './country-field-testing.module';
+
+import {
+  SkyCountryFieldFixture
+} from './country-field-fixture';
+
+const DATA_SKY_ID = 'test-country-field';
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/testing/country-field/country-field-fixture.spec.ts:29]: Check your alphabetical ordering of properties here.

File: src/app/public/testing/country-field/country-field-fixture.spec.ts [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+
+    fixture = TestBed.createComponent(
+      CountryFieldTestComponent
+    );
+    testComponent = fixture.componentInstance;
+    fixture.detectChanges();
+    countryFieldFixture = new SkyCountryFieldFixture(fixture, DATA_SKY_ID);
+  });
+
+  it('should expose the expected defaults', fakeAsync(async () => {
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/testing/country-field/country-field-fixture.spec.ts:91]: Do all of these tests need to be `async`? We probably also shouldn't be mixing `fakeAsync` and `async`.

File: src/app/public/testing/country-field/country-field-fixture.ts [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+   */
+  public get searchText(): string {
+    const inputEl = this.getInput();
+    return inputEl.nativeElement.value;
+  }
+
+  /**
+   * The value of the input field.
+   */
+  public get selectedCountryData(): SkyCountryFieldCountry {
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/testing/country-field/country-field-fixture.ts:80]: I'm wondering if this method is necessary? The `SkyCountryFieldComponent` has an output property `selectedCountryChange` that emits the selected country data - can the consumer just use that?

**Review by Blackbaud-AlexKingman — COMMENTED** [review]
Another great start! Left a few comments...

**Comment by blackbaud-ado** [comment]
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-country-test-fixture](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-country-test-fixture)

**Review by blackbaud-svossen — COMMENTED** [review]
@Blackbaud-AlexKingman All the comments should be taken care of. Let me know if there's anything else you find out of place!

File: src/app/public/testing/country-field/country-field-fixture.ts [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    private fixture: ComponentFixture<any>,
+    skyTestId: string
+  ) {
+    this.debugEl = SkyAppTestUtility.getDebugElementByTestId(fixture, skyTestId, 'sky-country-field');
+  }
+
+  /**
+   * The value of the input field's autocomplete attribute.
+   */
+  public get autocompleteAttribute(): string {
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/testing/country-field/country-field-fixture.ts:33]: Code style - These getters should be above the constructor.

**Review by Blackbaud-AlexKingman — COMMENTED** [review]
Just a few last comments/suggestions. This is looking great!

File: src/app/public/testing/country-field/country-field-fixture.ts [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+   */
+  public search(searchText: string): Promise<NodeListOf<HTMLElement>> {
+    return this.searchAndGetResults(searchText, this.fixture);
+  }
+
+  /**
+   * Enters the search text into the input field and selects the first result (if any).
+   * @param searchText The name of the country to select.
+   */
+  public select(searchText: string): Promise<any> {
```
[the remark this answers is not in the document]
Comment (blackbaud-svossen) [reply src/app/public/testing/country-field/country-field-fixture.ts:73]: I was copying the searchAndSelect private method from another unit test. I liked leaving that private method separate as it is not opinionated about which result is selected.

If you want I can certainly rename this wrapper/opinionated search method to searchAndSelect or even searchAndSelectFirstResult for clarity.

File: src/app/public/testing/country-field/country-field-fixture.spec.ts [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+    testComponent.autocompleteAttribute = 'on';
+    testComponent.disabled = true;
+    fixture.detectChanges();
+
+    // verify updated values
+    expect(countryFieldFixture.autocompleteAttribute).toBe(testComponent.autocompleteAttribute);
+    expect(countryFieldFixture.disabled).toBe(testComponent.disabled);
+  });
+
+  it('should expose selection properties', async () => {
```
[the remark this answers is not in the document]
Comment (blackbaud-svossen) [reply src/app/public/testing/country-field/country-field-fixture.spec.ts:97]: Sure, easy enough.

**Review by Blackbaud-AlexKingman — APPROVED** [review]

**Comment by blackbaud-ado** [comment]
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-country-test-fixture](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-country-test-fixture)

**Pull request merged**